File: Feiniubus/FNNetInterface/FNNetInterface/NetInterfaceManager.m
//
//  NetInterfaceManager.m
//  XinRanApp
//
//  Created by tianbo on 14-12-8.
//  Copyright (c) 2014年 feiniu. All rights reserved.
//

#import "NetInterfaceManager.h"
#import "NetInterFace.h"
#import "JsonBaseBuilder.h"
#import "UrlMaps.h"

#import <FNDataModule/FNDataModule.h>
#import <FNDataModule/EnvPreferences.h>
#import <FNCommon/Constants.h>
//#import "User.h"
#import <FNDataModule/NSObject+MJKeyValue.h>


typedef NS_ENUM(int, EmRequestMothed)
{
    RequestMothed_Put = 0,
    RequestMothed_Post,
    RequestMothed_Get,
    RequestMothed_Delete,
};


@interface NetInterfaceManager ()
{
    NSString *recordUrl;
    NSDictionary *recordBody;
    RequestType recordRequestType;
    int recordRequestMothed;
    
    NSString *controllerId;
}

@property(nonatomic, copy) void (^successBlock)(NSString* msg);
@property(nonatomic, copy) void (^failedBlock)(NSString* msg);

//url字典
@property (nonatomic, strong) UrlMaps *urpMaps;
@end

@implementation NetInterfaceManager

+(NetInterfaceManager*)sharedInstance
{
    static NetInterfaceManager *instance = nil;
    static dispatch_once_t once;
    dispatch_once(&once, ^{
        instance = [[self alloc] init];
    });
    
    return instance;
}


-(instancetype)init
{
    self = [super init];
    if (self) {
        _urpMaps = [[UrlMaps alloc]init];
    }
    return self;
}

-(void)dealloc
{
    self.successBlock = nil;
    self.failedBlock = nil;
}

/**
 *  设置鉴权字符串
 *
 *  @param auth 鉴权字符串
 */
-(void)setAuthorization:(NSString*)auth
{
    [[NetInterface sharedInstance] setAuthorization:auth];
}

- (int)reach
{
    return [[NetInterface sharedInstance] reach];
}

-(void)setSuccessBlock:(void (^)(NSString*))success failedBlock:(void (^)(NSString*))failed
{
    self.successBlock = success;
    self.failedBlock = failed;
}

#pragma mark-
- (void)uploadImage:(NSString*)strurl
                img:(UIImage*)img
               body:(NSDictionary*)body
       suceeseBlock:(void(^)(NSString* msg))suceese
        failedBlock:(void(^)(NSError* error))failed
{
    //绕过授权
//    strurl = [[NSString alloc ] initWithFormat:@"%@%@",strurl,@"?manager=xxx"];
    
    [[NetInterface sharedInstance] uploadImage:strurl body:body img:img suceeseBlock:^(NSString *msg) {
        suceese(msg);
    } failedBlock:^(NSError *error) {
        failed(error);
    }];
}


/**
 *  统一网络请求接口
 *
 *  @param type  请求类型
 *  @param block 参数配置回调
 */
- (void)sendRequstWithType:(int)type params:(void(^)(NetParams* params))block
{
    NetParams *params = [[NetParams alloc] init];
    if (block) {
        block(params);
    }
    
    NSString *url = [_urpMaps urlWithType:type];
    
    switch (params.method) {
        case EMRequstMethod_GET:
        {
            [self getRequst:url body:params.data requestType:type];
        }
            break;
        case EMRequstMethod_POST:
        {
            [self postRequst:url body:params.data requestType:type];
        }
            break;
        case EMRequstMethod_PUT:
        {
            [self putRequst:url body:params.data requestType:type];
        }
            break;
        case EMRequstMethod_DELETE:
        {
            [self deleteRequest:url body:params.data requestType:type];
        }
            break;
            
        default:
            break;
    }
    
}

#pragma mark-
-(void)postRequst:(NSString*)strUrl body:(NSDictionary*)body requestType:(int)type
{
    //绕过授权
//    strUrl = [[NSString alloc ] initWithFormat:@"%@%@",strUrl,@"?manager=xxx"];
    //记录上次请求
    [self recordRequst:strUrl body:body requestType:type mothed:RequestMothed_Post];
    
    __weak __typeof(self)weakSelf = self;
    [[NetInterface sharedInstance] httpPostRequest:strUrl body:body suceeseBlock:^(NSString *msg){
        
        __strong __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf successHander:msg reqType:type];
        
    }failedBlock:^(NSError *error) {
        
        __strong __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf failedHander:error reqType:type];
    }];
}

-(void)getRequst:(NSString*)strUrl body:(NSDictionary*)body requestType:(int)type
{
    //绕过授权
//    strUrl = [[NSString alloc ] initWithFormat:@"%@%@",strUrl,@"?manager=xxx"];
    //记录上次请求
    [self recordRequst:strUrl body:body requestType:type mothed:RequestMothed_Get];
    
    __weak __typeof(self)weakSelf = self;
    [[NetInterface sharedInstance] httpGetRequest:strUrl body:body suceeseBlock:^(NSString *msg){
        
        __strong __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf successHander:msg reqType:type];
        
    }failedBlock:^(NSError *error) {
        
        __strong __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf failedHander:error reqType:type];
        
    }];
}

-(void)putRequst:(NSString*)strUrl body:(NSDictionary*)body requestType:(int)type
{
    //记录上次请求
    [self recordRequst:strUrl body:body requestType:type mothed:RequestMothed_Put];
    
    __weak __typeof(self)weakSelf = self;
    [[NetInterface sharedInstance] httpPutRequest:strUrl body:body suceeseBlock:^(NSString *msg){
        
        __strong __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf successHander:msg reqType:type];
        
    }failedBlock:^(NSError *error) {
        
        __strong __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf failedHander:error reqType:type];
        
    }];
}
- (void)deleteRequest:(NSString *)strUrl body:(NSDictionary*)body requestType:(int)type{
    [self recordRequst:strUrl body:body requestType:type mothed:RequestMothed_Delete];
    
    __weak __typeof(self)weakSelf = self;
    [[NetInterface sharedInstance] httpDeleteRequest:strUrl body:body suceeseBlock:^(NSString *msg) {
        __strong __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf successHander:msg reqType:type];
    } failedBlock:^(NSError *error) {
        __strong __typeof(weakSelf)strongSelf = weakSelf;
        [strongSelf failedHander:error reqType:type];
    }];
}

-(void)successHander:(NSString*)msg reqType:(int)reqestType
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        
        NSDictionary *dict = [JsonBaseBuilder dictionaryWithJson:msg decode:NO key:nil];
        if (dict) {
            if ([dict objectForKey:@"code"] &&
                ([[dict objectForKey:@"code"] intValue] == 100002 ||
                 [[dict objectForKey:@"code"] intValue] == 100001) &&
                reqestType!= KRequestType_Login)         //后台鉴权失败
            {
                dispatch_async(dispatch_get_main_queue(), ^{
                    NSError *error = [NSError errorWithDomain:@"" code:401 userInfo:nil];
                    NSMutableDictionary *result = [@{@"type": [NSNumber numberWithInt:reqestType], @"error":error} mutableCopy];
                    
                    [[NSNotificationCenter defaultCenter] postNotificationName:KNotification_RequestFailed object:result];
                });
            }
            else {
                dispatch_async(dispatch_get_main_queue(), ^{
                    
                    ResultDataModel *obj = [[ResultDataModel alloc] initWithDictionary:dict reqType:reqestType];
                    
                    //通知页面
                    [[NSNotificationCenter defaultCenter] postNotificationName:KNotification_RequestFinished object:obj];
                });
            }
        }
        else {
            DBG_MSG(@"dict is nil!!!");
            dispatch_async(dispatch_get_main_queue(), ^{
                NSError *error = [NSError errorWithDomain:@"" code:405 userInfo:@{NSLocalizedDescriptionKey:@"数据异常"}];

                NSMutableDictionary *result = [@{@"type": [NSNumber numberWithInt:reqestType], @"error":error} mutableCopy];

                [[NSNotificationCenter defaultCenter] postNotificationName:KNotification_RequestFailed object:result];
            });

        }
    });
}

-(void)failedHander:(NSError*)error reqType:(int)reqestType
{
    dispatch_async(dispatch_get_main_queue(), ^{
        //通知页面
        NSMutableDictionary *result = [@{@"type": [NSNumber numberWithInt:reqestType], @"error":error} mutableCopy];
        
        [[NSNotificationCenter defaultCenter] postNotificationName:KNotification_RequestFailed object:result];
    });
    
//    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
//        //转化错误信息到ResultDataModel
////        ResultDataModel *result = [[ResultDataModel alloc] initWithErrorInfo:error reqType:reqestType];
//        
////        if (result.resultCode == 401 || result.resultCode == 403) {
////            //鉴权失效重置token
////            [[EnvPreferences sharedInstance] setToken:nil];
////            [[EnvPreferences sharedInstance] setUserId:nil];
////            //[self tokenExpireHandler];
////            return ;
////        }
//        
//        dispatch_async(dispatch_get_main_queue(), ^{
//            //通知页面
//            NSDictionary *result = @{@"type": [NSNumber numberWithInt:reqestType], @"error":error};
//            [[NSNotificationCenter defaultCenter] postNotificationName:KNotification_RequestFailed object:result];
//        });
//    });
}

#pragma mark-

//重新加载上一次的请求
-(void) reloadRecordData {
    
    if (recordUrl && recordBody) {
        switch (recordRequestMothed) {
            case RequestMothed_Post:
                [self postRequst:recordUrl body:recordBody requestType:recordRequestType];
                break;
            case RequestMothed_Get:
                [self getRequst:recordUrl body:recordBody requestType:recordRequestType];
                break;
            case RequestMothed_Put:
                [self putRequst:recordUrl body:recordBody requestType:recordRequestType];
                break;
            case RequestMothed_Delete:
                [self deleteRequest:recordUrl body:recordBody requestType:recordRequestType];
                break;
            default:
                
                break;
        }
    }
    
}

-(void) recordRequst:(NSString*)strUrl body:(NSDictionary*)body requestType:(int)type mothed:(EmRequestMothed)mothed {
    //记录一次请求数据
    recordUrl           = strUrl;
    recordBody          = body;
    recordRequestType   = type;
    recordRequestMothed = mothed;
}

//会话过期处理方法
//-(void)tokenExpireHandler {
//
//    //// stup 1. 登录
//    User *user = [[EnvPreferences sharedInstance] getUserInfo];
//    if (user && user.username && user.pwd) {
//
//        NSString *body = [[JsonBuilder sharedInstance] jsonWithLogin:user.username pwd:user.pwd type: LoginType_Email];
//        NSString *url = [NSString stringWithFormat:@"%@%@", KServerAddr, KUrl_Login];
//
//        [[NetInterface sharedInstance] httpPostRequest:url body:body suceeseBlock:^(NSString *msg){
//
//            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
//
//                //转化josn数据到ResultDataModel
//                ResultDataModel *result = [[ResultDataModel alloc] initWithDictionary:[JsonBuilder dictionaryWithJson:msg decode:NO key:nil] reqType:KReqestType_Login];
//
//                if (result.resultCode == 0) {
//                    //解析出token字符串
//                    NSString *token = [JsonBuilder tokenStringWithJosn:msg decode:NO key:nil];
//                    [[EnvPreferences sharedInstance] setToken:token];
//
//                    //// stup 2. 重新请求
//                    [self reloadRecordData];
//                }
//                else {
//                    dispatch_async(dispatch_get_main_queue(), ^{
//                        //通知页面
//                        [[NSNotificationCenter defaultCenter] postNotificationName:KNotification_RequestFailed object:result];
//                    });
//                }
//
//            });
//
//        }failedBlock:^(NSError *error) {
//
//            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
//                //转化错误信息到ResultDataModel
//                ResultDataModel *result = [[ResultDataModel alloc] initWithErrorInfo:error reqType:KReqestType_Login];
//
//                dispatch_async(dispatch_get_main_queue(), ^{
//                    //通知页面
//                    [[NSNotificationCenter defaultCenter] postNotificationName:KNotification_RequestFailed object:result];
//                });
//
//            });
//
//        }];
//    }
//}



/**
 *  设置当前的ViewController
 *
 *  @param controller name
 */
-(void)setReqControllerId:(NSString*)cId
{
    //DBG_MSG(@"--controllerid=%@", cId)
    controllerId= cId;
}

-(NSString*)getReqControllerId
{
    return controllerId;
}





///**
// *  微信支付接口
// *
// *  @param pdtName  商品名称
// *  @param price 总价
// *  @param orderId  订单id
// */
//-(void)weixinPay:(NSString*)pdtName price:(NSString*)price orderId:(NSString*)orderId
//{
//    NSString *body = [[JsonBuilder sharedInstance] jsonWithweixinPay:pdtName price:price orderId:orderId];
//    NSString *url = [NSString stringWithFormat:@"%@%@", KServerAddr, Kurl_WXPay];
//
//    [self postRequst:url body:body requestType:KReqestType_weixinPay];
//}
//
///**
// *  微信支付确认接口
// *
// *  @param orderId 订单id
// */
//-(void)checkWXPay:(NSString*)orderId
//{
//    NSString *body = [[JsonBuilder sharedInstance] jsonWithCheckWxPay:orderId];
//    NSString *url = [NSString stringWithFormat:@"%@%@", KServerAddr, Kurl_Checkwxpay];
//
//    [self postRequst:url body:body requestType:KReqestType_CheckWXPay];
//}
///**
// *  支付宝支付接口
// *
// *  @param pdtName  商品名称
// *  @param price 总价
// *  @param orderId  订单id
// */
//-(void)aliPay:(NSString*)pdtName price:(NSString*)price orderId:(NSString*)orderId
//{
//    NSString *body = [[JsonBuilder sharedInstance] jsonWithweixinPay:pdtName price:price orderId:orderId];
//    NSString *url = [NSString stringWithFormat:@"%@%@", KServerAddr, Kurl_AliPay];
//    
//    [self postRequst:url body:body requestType:KReqestType_AliPay];
//}


@end

File: Feiniubus2.0/FNSwiftNetInterface/FNSwiftNetInterface/NetInterfaceManager.swift
//
//  NetInterfaceManager.swift
//  FNNetInterface
//
//  Created by lbj@feiniubus.com on 16/8/25.
//  Copyright © 2016年 FN. All rights reserved.
//

import UIKit

public let KNotification_RequestFinished:String = "HttpRequestFinishedNoitfication"
public let KNotification_RequestFailed:String = "HttpRequestFailedNoitfication"
public let KNotification_AuthenticationFail:String =  "AuthenticationLogicFailNoitfication"



open class NetInterfaceManager: NSObject {
    
    //单例
    open static let shareInstance = NetInterfaceManager()
    fileprivate override init() { super.init() }
    
    fileprivate var recordUrl:String?
    fileprivate var recordBody:NSDictionary?
    fileprivate var recordRequestType:Int = 0
    fileprivate var recordRequestMothed:EMRequstMethod = .emRequstMethod_GET
    fileprivate var controllerId:String?
    
    open func setAuthorization(_ httpHeader:[String:String]) -> Void{
        NetInterface.shareInstance.setAuthorization(httpHeader)
    }
    
    open func getAuthorization() ->[String:String]{
        return NetInterface.shareInstance.httpHeader;
    }
    
    //MARK: Request
    open func sendRequstWithType(_ type:Int,block:((_ params:NetParams)->Void)) -> Void{
        
        let params:NetParams = NetParams.init()
        block(params);
//        guard let url:String? = UrlMaps.shareInstance.urlWithTypeNew(type) else {return}
        let url:String? = UrlMaps.shareInstance.urlWithTypeNew(type)
        
        let bodyDic:NSDictionary? = params.data as? NSDictionary
        NetInterface.shareInstance.httpRequest(params.method!, strUrl:url!, body:bodyDic as? [String:AnyObject], successBlock: { (msg) in
            self.successHander(msg, reqestType: type)
            }, failedBlock: { (msg, error) in
                self.failedHander(msg, error: error, reqestType: type)
        })
        
        
    }
    
    open func sendFormRequstWithType(_ type:Int, block:((_ params:NetParams)->Void)) -> Void{
        
        let params:NetParams = NetParams.init()
        block(params);
//        guard let url:String? = UrlMaps.shareInstance.urlWithTypeNew(type) else {return}
        let url:String? = UrlMaps.shareInstance.urlWithTypeNew(type)
        
        let bodyDic:NSDictionary? = params.data as? NSDictionary
        NetInterface.shareInstance.httpFormRequest(EMRequstMethod.emRequstMethod_POST, strUrl:url!, body:bodyDic as? [String:AnyObject], successBlock: { (msg) in
                self.successHander(msg, reqestType: type)
            }, failedBlock: { (msg, error) in
                self.failedHander(msg, error: error, reqestType: type)
            })
    
    }
    //MARK: Private
    //返回成功处理
    fileprivate func successHander(_ msg:String, reqestType:Int) ->Void{
        let data:Data = msg.data(using: String.Encoding.utf8)!
        var dict:AnyObject?
        if data.count > 0 {
            do{
                dict = try JSONSerialization.jsonObject(with: data, options:JSONSerialization.ReadingOptions.allowFragments) as AnyObject
            }
            catch{
                dict = nil
            }
        }
        
        let result:NetResultDataModel = NetResultDataModel.initWithDictionary(dict!, reqestType: reqestType)
        if result.code == EMResultCode.emCode_Success.rawValue {
            DispatchQueue.main.async {
                NotificationCenter.default.post(name: Notification.Name(rawValue: KNotification_RequestFinished), object: result)
            }
        }
        else{
            DispatchQueue.main.async {
                NotificationCenter.default.post(name: Notification.Name(rawValue: KNotification_RequestFailed), object: result)
            }
        }
    
    }
    //返回失败处理
    fileprivate func failedHander(_ msg:String, error:NSError, reqestType:Int) ->Void{
        let data:Data = msg.data(using: String.Encoding.utf8)!
        var dict:AnyObject?
        if data.count > 0 {
            do{
                dict = try JSONSerialization.jsonObject(with: data, options:JSONSerialization.ReadingOptions.allowFragments) as AnyObject
            }
            catch{
                dict = nil
            }
        }
        
        
        DispatchQueue.main.async {
            let result:NetResultDataModel = NetResultDataModel.initWithErrorInfo(dict, error: error, reqestType: reqestType)
            if result.code == EMResultCode.emCode_TokenOverdue.rawValue{
                NotificationCenter.default.post(name: Notification.Name(rawValue: KNotification_AuthenticationFail), object: result)
            }
            else{
                NotificationCenter.default.post(name: Notification.Name(rawValue: KNotification_RequestFailed), object: result)
            }
        }
        
    }
    
    
    //MARK: recod
    fileprivate func recodRequest(_ strUrl:String, body:NSDictionary, requestType:Int, mothed:EMRequstMethod){
        //记录一次请求数据
        recordUrl           = strUrl;
        recordBody          = body;
        recordRequestType   = requestType;
        recordRequestMothed = mothed;
    }
    
    open func reloadRecordData()->Void{
        if (recordUrl != nil && recordBody != nil) {
            NetInterface.shareInstance.httpFormRequest(recordRequestMothed, strUrl:recordUrl!, body:recordBody! as! [String : AnyObject], successBlock: { (msg) in
                    self.successHander(msg, reqestType:self.recordRequestType)
                }, failedBlock: { (msg, error) in
                    self.failedHander(msg, error: error, reqestType:self.recordRequestType)
            })
        }
    }
    
    open func setReqControllerId(_ cId:String) -> Void{
        controllerId = cId;
    }
    
    open func getReqControllerId()->String{
        return controllerId!;
    }
}











